### ISSUE TITLE
Add to FreeCAD Addon Manager

### ISSUE THREAD

**Issue by luzpaz** [ebrahimraeyat/OSAFE#27]
https://github.com/FreeCAD/FreeCAD-addons

**Comment by ebrahimraeyat** [comment]
Thanks. this page did not open:

[FreeCAD-Addon_Details](https://github.com/FreeCAD/FreeCAD-addons/blob/master/FreeCAD-Addon-Details.md)

**Comment by luzpaz** [comment]
Yea, it was removed in https://github.com/FreeCAD/FreeCAD-addons/issues/145

**Comment by ebrahimraeyat** [comment]
Hi @luzpaz , as my SteelColumn WB added to Addons, can you add this WB to FreeCAD-addons? thanks.

**Comment by luzpaz** [comment]
This also need a `metadata` file that includes external and internal dependencies.

**Comment by ebrahimraeyat** [comment]
yes, I added it for SteelColumn, I will add it. thanks

**Comment by luzpaz** [comment]
Heads up that we have a new updated metadata schema for the Addon Manager. See https://wiki.freecad.org/Package_Metadata

**Comment by ebrahimraeyat** [comment]
Yes, I saw it, example:
https://github.com/yorikvanhavre/BIM_Workbench/pull/86

**Comment by ebrahimraeyat** [comment]
question:
my OSAFE WB needed [etabs_api](https://github.com/ebrahimraeyat/etabs_api) module. How can I add this to depend in package.xml?

<content>
    <workbench>
      <classname>OSAFE</classname>
      <subdirectory>./</subdirectory>
      <depend kind="freecad">Arch</depend>
    </workbench>
  </content>

**Comment by luzpaz** [comment]
@chennes care to weigh-in?

**Comment by chennes** [comment]
The only types of dependency information FreeCAD can use are Python packages installable via pip, or other FreeCAD Addons. Other code dependencies should either be submodules, or just be specified in your documentation.

**Comment by ebrahimraeyat** [comment]
> The only types of dependency information FreeCAD can use are Python packages installable via pip, or other FreeCAD Addons. Other code dependencies should either be submodules, or just be specified in your documentation.

Thanks. pip can install from github, is it possible to add this option to freecad?

**Comment by chennes** [comment]
Not really -- I mean, yes, technically it's "possible", but it's really just using pip as a download manager. Using a submodule is more appropriate in this case, IMO.

**Comment by ebrahimraeyat** [comment]
I had been used submodule for a while, but it has a problem for the first time that user wanted to update it, It did not download submodule and user must do some commands manually and unfortunately my users couldn't do that. I also tried to automate everythings, but not succeed.

**Comment by chennes** [comment]
That will be automated by the Addon Manager if they install that way -- it handles submodules.

**Comment by ebrahimraeyat** [comment]
you mean Addon Manager can automatically install submodules, even in first time? or user must do it manually for the first time? Thanks.

**Comment by chennes** [comment]
> you mean Addon Manager can automatically install submodules, even in first time

Yes, the Addon Manager should do it automatically-- actually that isn't new, it was part of the original code before I started. If it is not working let me know.

**Comment by ebrahimraeyat** [comment]
I will check and tell you.

**Comment by luzpaz** [comment]
ping

**Comment by ebrahimraeyat** [comment]
> > you mean Addon Manager can automatically install submodules, even in first time
> 
> Yes, the Addon Manager should do it automatically-- actually that isn't new, it was part of the original code before I started. If it is not working let me know.

I tested it and it works. but one question:
when I change my submodule and add & commit & push on github, my main repo still show modified in submodule, also I must add & commit & push in my main repo?

**Comment by chennes** [comment]
The way git submodules work, you are pinned to a specific revision, so yes, you also have to update your main repo. That's not an Addon Manager thing, that's a git thing.

**Comment by ebrahimraeyat** [comment]
yes, I know, I wanted to be sure. probably I mention it in README file that user must install it manually. Thanks

**Comment by luzpaz** [comment]
Any progress on this, @ebrahimraeyat ?

**Comment by chennes** [comment]
The changed git code doesn't affect anything related to pip.

**Comment by ebrahimraeyat** [comment]
Thanks @chennes . I added automatic update if user confirms the update with https://github.com/ebrahimraeyat/etabs_api/commit/4a28fe5b62a8334521b15435ac73620116eda79b

**Comment by luzpaz** [comment]
Ok, what's left here to do ?

**Comment by ebrahimraeyat** [comment]
@luzpaz . Do you mean I must do something else?

**Comment by luzpaz** [comment]
I haven't tested but can OSAFE be added via the Addon Manager?

**Comment by ebrahimraeyat** [comment]
> I haven't tested but can OSAFE be added via the Addon Manager?

Yes, But it has some dependencies like Pandas, comtypes and maybe some additional.  I have created a version with all dependencies:

https://github.com/ebrahimraeyat/OSAFE/releases/tag/v0.9

**Comment by chennes** [comment]
We do support adding dependencies automatically: you can specify them in package.xml and the Addon Manager will offer to install them for the user if they are needed. We have a "safe list" of packages that we'll automatically install, so we may need to expand that if you need things that aren't on it.

**Comment by ebrahimraeyat** [comment]
Dependencies That I found not in `safe list`:
- comtypes
- pydocx[optional]

**Comment by chennes** [comment]
I added com types and pydocx, but etabs_api isn't in pypi (that I could find), so the Addon Manager won't be able to install that one automatically.

**Comment by chennes** [comment]
Also, here's a PR to actually add those two dependencies:
https://github.com/ebrahimraeyat/OSAFE/pull/37

**Comment by ebrahimraeyat** [comment]
> I added com types and pydocx, but etabs_api isn't in pypi (that I could find), so the Addon Manager won't be able to install that one automatically.

Thanks. Actually etabs_api needs comtypes. I will see if I could create a pip version for etabs_api

**Comment by luzpaz** [comment]
@ebrahimraeyat please update this ticket when you do, thanks!

**Comment by ebrahimraeyat** [comment]
@luzpaz, updated  :
[etabs_api](https://pypi.org/project/etabs-api/1.0/)

**Comment by luzpaz** [comment]
:point_up_2:

**Comment by ebrahimraeyat** [comment]
Added etabs-api as dependency: https://github.com/ebrahimraeyat/OSAFE/commit/f095e08aa686a7889b455f162854c3dffe025d0d

**Comment by ebrahimraeyat** [comment]
> Where are we at ?

I think nothing more job to do here. All things done, am I right?

**Comment by luzpaz** [comment]
Can you summarize what has been done?

**Comment by ebrahimraeyat** [comment]
> Can you summarize what has been done?

Thanks for your reply.
The main problem was the etabs_api as an external dependency. I created a pypi package for it and you added etabs-api as an  allowed python package to be install via Addon manager. I also added it as an dependency in package.xml of OSAFE.

**Comment by luzpaz** [comment]
Thanks!  
Feel free to submit a PR to https://github.com/FreeCAD/FreeCAD-addons to include in the Addon Manager

**Comment by luzpaz** [comment]
ready to do this?

**Comment by ebrahimraeyat** [comment]
Thanks @luzpaz . Yes, I will do it.

**Comment by luzpaz** [comment]
Good job!
